js/src/forum/components/Search.js

```diff
@@ -217,7 +214,7 @@ export default class Search extends Component {
     this.value('');
 
     if (this.getCurrentSearch()) {
-      app.current.clearSearch();
+      app.search.clearSearch();
```

Comment:
This should work as well (while being more consistent), right?

```suggestion
      this.state.clearSearch();
```

Reply:
Tbh, I think it might be better to move the entire method into SearchState, along with the value prop.

Reply:
Including the Mithril stuff (the prop and the `redraw()`)? Do we want rendering / framework stuff in those states?

Reply:
This is a difficult question, but I'm leaning towards yes on this one. My strictest no-no would be including any jsx in state methods (not the biggest fan of storing JSX in states as in #2163, but that's not AS critical). The way I see it, every state method is either involved in:

- Changing state data
- Reacting to changes in state data by setting off animations / redraws.
- Presenting the public API for components to interact with other components (also bad, at no point should any component have access to an instance of any other component (although we can use events/pass bound functions to circumvent some of the limitations))

So redraws are fine, as is direct DOM manipulation via JQuery (although when we CAN, I'd prefer to still have that in component classes, especially if it's a big method).

Reply:
One other thing I'll note is that I view states as an intermediate step on the way to a proper state management methodology, at which point the data in states and the public API of manipulating state would be further split off. But one step at a time.

Reply:
Don't read this / reply here, I posted an expanded version in the main 'Frontend Needs Better State' discussion

Reply:
I will comment on the specific issue I commented on. I can live with redraws in states for now, but would prefer trying to avoid props in there. (Then again, these props are just setters that trigger redraws, if I understand correctly?)

Reply:
I moved value into state (and de-prop-ized it), but am going to keep clear() in the component, as Byobu needs to override that. Besides, it has to do with component functionality, not global search functionality.

Reply:
Just to quickly expand on this confusing wording:

- the `value` attribute of `Search.js` has been moved into `SearchState`
- Since a mithril prop is no longer necessary for it to function properly, it now uses a regular old attribute.
- Byobu overrides the functionality of the `clear` method in https://github.com/FriendsOfFlarum/byobu/blob/6ad1252790aa01329c01d6a5d2ed3c458b32cc67/js/src/forum/components/RecipientSearch.js#L129. If we want to keep this flexibility without encouraging people to make classes extending SeachState, `clear` needs to be in the Search.js class.